src/hashtable.c

```diff
@@ -0,0 +1,163 @@
+/*= -*- c-basic-offset: 4; indent-tabs-mode: nil; -*-
+ *
+ * hashtable.c -- a generic hashtable implementation.
+ *
+ * Copyright (C) 2016 by Donovan Baarda <abo@minkirri.apana.org.au>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
+ */
+#include <assert.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include "hashtable.h"
+
+/* Open addressing works best if it can take advantage of memory caches using
+ * locality for probes of adjacent buckets on collisions. So we pack the keys
+ * tightly together in their own key table and avoid referencing the element
+ * table and elements as much as possible. Key value zero is reserved as a
+ * marker for an empty bucket to avoid checking for NULL in the element table.
+ * If we do get a hash value of zero, we -1 to wrap it around to 0xffff. */
+
+/* Use max 0.8 load factor to avoid bad open addressing performance. */
+#define HASHTABLE_LOADFACTOR_NUM 8
+#define HASHTABLE_LOADFACTOR_DEN 10
```

Comment:
Yes, but with the tight packing of the hash values it doesn't hurt that much. For files with size/loadfactors of 512K/0.5, 716K/0.7, 819K/0.8 1024K/0.5 we get;

512K/0.5
=======
librsync: match statistics: signature[268697600 searches, 262144 (0.098%) matches, 313791392 (1.168x) weak sum compares, 518749 (0.193%) strong sum compares, 518434 (0.193%) strong sum calcs]
librsync: delta statistics: literal[8192 cmds, 268435456 bytes, 24576 cmdbytes] copy[2 cmds, 268435456 bytes, 0 false, 18 cmdbytes] speed[536.9 MB (7.5 MB/s) in, 268.5 MB (3.7 MB/s) out, 72 sec]

716K/0.7
=======
librsync: match statistics: signature[375756800 searches, 366592 (0.098%) matches, 1053222017 (2.803x) weak sum compares, 867330 (0.231%) strong sum compares, 866681 (0.231%) strong sum calcs]
librsync: delta statistics: literal[11456 cmds, 375390208 bytes, 34368 cmdbytes] copy[2 cmds, 375390208 bytes, 0 false, 18 cmdbytes] speed[750.8 MB (7.0 MB/s) in, 375.4 MB (3.5 MB/s) out, 108 sec]

819K/0.8
=======
librsync: match statistics: signature[429811200 searches, 419328 (0.098%) matches, 2103451089 (4.894x) weak sum compares, 1075564 (0.250%) strong sum compares, 1074655 (0.250%) strong sum calcs]
librsync: delta statistics: literal[13104 cmds, 429391872 bytes, 39312 cmdbytes] copy[2 cmds, 429391872 bytes, 0 false, 18 cmdbytes] speed[858.8 MB (6.2 MB/s) in, 429.4 MB (3.1 MB/s) out, 139 sec]

1024K/0.5
=======
librsync: match statistics: signature[537395200 searches, 524288 (0.098%) matches, 627469786 (1.168x) weak sum compares, 1548850 (0.288%) strong sum compares, 1547240 (0.288%) strong sum calcs]
librsync: delta statistics: literal[16384 cmds, 536870912 bytes, 49152 cmdbytes] copy[2 cmds, 536870912 bytes, 0 false, 18 cmdbytes] speed[1073.7 MB (7.5 MB/s) in, 536.9 MB (3.8 MB/s) out, 143 sec]

You can see that at loadfactor 0.8 it did get as high as averaging 4.894 hash compares per search and the execution time of 139secs was a fair bit worse than the 108secs the 716K/0.7 took, but the execution time was still better than the 143secs the 1024K/0.5 one took.

If I drop the max loadfactor to 0.7 the the match and delta stats become;

819K/0.4
=======
librsync: match statistics: signature[429811200 searches, 419328 (0.098%) matches, 327713444 (0.762x) weak sum compares, 1075564 (0.250%) strong sum compares, 1074655 (0.250%) strong sum calcs]
librsync: delta statistics: literal[13104 cmds, 429391872 bytes, 39312 cmdbytes] copy[2 cmds, 429391872 bytes, 0 false, 18 cmdbytes] speed[858.8 MB (8.3 MB/s) in, 429.4 MB (4.2 MB/s) out, 103 sec]

The execution time drops from 139secs to 103secs, which is a pretty good ~25% improvement, but not that astounding and at the cost of using significantly more memory. I'm happy to change it if people prefer the tradeoff for more speed though.